discussions/micro-frontends/base-paths.md

```diff
@@ -0,0 +1,160 @@
+
+We can configure the build tooling to prefix all asset paths in generated HTML and JavaScript.
+For example:
+
+```ts
+export default defineConfig({
+  ...
+  base: "/_fragment/todos/",
+  ...
+});
+```
+
+This is effective for ensuring that generated client side JS code is correctly prefixed; but it can miss asset paths that the build tooling is not aware of.
+The build tooling can be made aware of asset paths if the developer ensures that they are accessed via ESM imports.
+For example:
+
+```ts
+import "./App.css";
+```
+
+Paths to assets that are not known to the build tooling, such as image `src` tags, or dynamically generated import paths can be missed.
+
+This approach does not allow for the asset path to be determined at runtime based on how a fragment is being used.
+Since the path to assets may be different in each case, this can cause the fragment to work in only standalone or within the context of a container, but not both.
+
+One solution to this is to actually host the assets at a prefixed path even when the fragment is standalone.
+If the assets are always hosted at `/_fragment/todos/...` then the fragment can always reference these consistently wherever it is running.
+This requires the prefix is unique across all container applications that use the fragment.
+
+### Dynamic base path prefixing
+
+Some frameworks, such as Qwik, allow the "base path" to be configured at runtime, when server-side rendering the HTML.
+This approach is equivalent to the build-time configuration, but gives more flexibility to change the base path depending upon usage of the fragment.
+When requesting a fragment's HTML a parameter could be added to the request specifying the asset path, which can be used when server-side rendering the HTML.
+For example:
+
+```ts
+	const url = new URL(request.url);
+	const fragmentBase = url.searchParams.get("base") ?? "/";
+	renderToStream(RootNode, {
+		...
+		envData: { request, env, fragmentBase },
+		base: fragmentBase + "build/",
+	})
+```
+
+Here the Qwik `renderToStream()` function accepts a `base` property which it uses to ensure that paths to client-side JavaScript files are correct.
+In the above example the `fragmentBase` is also added to the `envData`, which is then available via the `useEnvData()` hook in components to use when rendering asset paths in templates.
+
+### Magic base-path token
+
+If the framework does not provide a mechanism to resolve the base path at runtime, it might be possible to use a "magic token" (a unique string) for the base path to assets.
+This string will appear in the built JavaScript and the rendered HTML.
+The magic token can be replaced with the appropriate path at runtime on the server (in a Worker) when it is found in the generated HTML or JavaScript when it is served.
+
+## What next?
+
+As we discuss these approaches with application developers and framework maintainers, we hope to converge on one or more common approach to solving these problems.
```

Comment:
```suggestion
As we discuss these approaches with application developers and framework maintainers, we hope to converge on one or more common approaches to solving these problems.
```

Reply:
OMG - I really didn't realise that was wrong until you pointed it out and I thought about it harder. Thanks.